### ISSUE TITLE
Use DL4TO for a 2D TO problem

### ISSUE THREAD

**Issue by acse-tz1023** [dl4to/dl4to#4]
Dear DL4TO team,

This is Terry, and I wish to use the DL4TO module to solve a 2D Topology Optimization problem. Is there a quick way to apply this module (especially the solver) to a 2D TO problem? I've tried to set the shape to something like [40, 4] (drop the z dimension) and [40, 4, 1] (set a dummy dimension of 1), but neither of them works. It seems the plotting AND solving part will both fail. 

Additionally, I am not sure why a pyvista 0.38.1 won't work on my end, but version 0.36.1 works well.

Best,
Terry

**Comment by dl4to** [comment]
Dear Terry,
it is possible to work on 2D problems with the DL4TO library, however the library is not intended that way. Therefore I cannot guarantee that the solution from the PDE solver will actually be correct in these cases. You indeed run into problems with the plotting and FDM module if you define shapes with a dummy dimension of 1. However, what you can do is you can use padding (that means, that you set the dimension to 3 and tread the first and third of these channels as basically empty channels). For instance, you could setup a problem like this (this is an adapted version of the problem that we define in the first tutorial on "problems and solutions":

```
import torch

# scalar material properties
E = 7e10 # Young's modulus (in Pascal)
ν = .3 # Poisson's ratio
σ_ys = 4.5e8 # Yield stress (in Pascal)

shape = [40, 4, 3] # number of voxels in x, y and z direction (we use 3 channels in z-direction for padding)
h = [0.0250, 0.0250, 0.0250] # spacial dimensions of each voxel in x, y and direction (in meters)

# define external forces
F = torch.zeros(3, *shape)
F[-1, :, :, 1] = -4e7 # external forces (in Newton)

# define locations of homogeneous Dirichlet conditions
Ω_dirichlet = torch.zeros(3, *shape)
Ω_dirichlet[:, 0, :, 1] = 1

# define design space
Ω_design = -torch.ones(1, *shape) # -1s indicate that these voxels can be freely optimized
Ω_design[:,:2,:,1] = 1. # we set densities to 1 where there are Dirichlet boundary conditions
Ω_design[:,:,:,1] = 1. # we set densities to 1 where there are external forces
```

As I said, if you do it like this, you will be able to execute the tutorial code without running into errors. However, I cannot fully guarantee that the PDE solution will be correct since I did not do specific testing for these cases. Also, since in our library the boundary conditions (for instance, the external forces) are averaged over the voxel volume, the FDM solver might be confused on which boundary/"side of the voxel" the forces are applied, if they are applied to a structure that is only one voxel thick.

Best,
David

**Comment by dl4to** [comment]
By the way, as an alternative, the FDM solver also has an in-built padding option: You simply define your FDM like this:
```
from dl4to.pde import FDM

problem.pde_solver = FDM(padding_depth=1)
```
Then, it is also possible to solve the PDE for problems with a dummy dimension of 1. What padding_depth=1 means is that for the solution of the PDE, the problem is being padded with an extra channel on each side and in each dimension. This of course makes the PDE solver much slower since the problem size is artificially enlarged. I would recommend the above first option though, since then also the plotting works.

**Comment by acse-tz1023** [comment]
A follow-up question to your first reply:
From the slightly modified version to the code in tutorial1, I am not sure why you said " I cannot fully guarantee that the PDE solution". In my opinion, after the padding to the z-direction, the problem will become an ordinary 3D problem and thus should perfectly compatible with the DL4TO module and all functionalities. 
Is there any possibility that the module and built-in solvers would fail for a shape [40, 4, 3]?

Thanks for the swift and thorough reply!

**Comment by DavidErzmann** [comment]
yes, but in our implementation we project boundary conditions that are defined on the surfaces of voxels (like surface forces) onto the voxels themselves, i.e., we approximate surface forces as volume forces over the voxel elements. This has the advantage that all information is defined voxel-wise, and the tensors have all the same size. However, if the structure is only one voxel thick (even if there are empty dummy voxels around), then this averaging operation loses the information on which side of the structure the force has been applied exactly (so it is not clear if the force is pulling on one side of the voxel or pushing on the other side). This might lead to inaccuracies in the solution and as I said, I did not specifically test for this case. Maybe it works though, feel free to try!
